functions/server/fn_startRecord.sqf

```diff
@@ -12,124 +14,106 @@ diag_log format ["grad replay: starting record with precision %1", _precision];
-
-    // only skip if currently paused
-    if (!GRAD_REPLAY_RECORDING_PAUSED) then {
-
-	    	_players = playableUnits + switchableUnits + allDeadMen;
-	    	_vehicles = vehicles + allDead - allDeadMen;
-	    	_ai = allUnits - playableUnits - switchableUnits;
-	    	_trackedUnits = _players;
-	    	_terminals = allMissionObjects "Land_DataTerminal_01_F"; // speciality for mission "breaking contact"
-	    	_trackedUnits append _terminals;
-
-	    	if (GRAD_REPLAY_AI_VEHICLES_TRACKED) then {
-	    		_trackedUnits append _vehicles;
-	    	};
+    _args params ["_currentSaveState"];
 
-	    	if (GRAD_REPLAY_AI_ONFOOT_TRACKED) then {
-	    		_trackedUnits append _ai;
-	    	};
-
-	    	// diag_log format ["tracked: %1",_trackedUnits];
+	// end recording and start playback
+	if (GRAD_REPLAY_RECORDING_STOPPED) exitWith {
+		[_handle] call CBA_fnc_removePerFrameHandler;
+		[] spawn GRAD_replay_fnc_preparePlaybackServer;
+	};
 
-		    {
-		    	_unit = _x;
+    // skip if currently paused
+    if (GRAD_REPLAY_RECORDING_PAUSED) exitWith {};
 
+	_players = playableUnits + switchableUnits + allDeadMen;
+	_trackedUnits = _players;
+	_terminals = allMissionObjects "Land_DataTerminal_01_F"; // speciality for mission "breaking contact"
+	_trackedUnits append _terminals;
 
-		    	// if (_unit getVariable ["GRAD_replay_track", false] || isPlayer _unit) then {
+	if (GRAD_REPLAY_VEHICLES_TRACKED) then {
+		// filter weapon holders
+		_vehicles = (vehicles + allDead - allDeadMen) select {
+			!(typeOf _x in ["WeaponHolder","WeaponHolder_Single_F","WeaponHolderSimulated","WeaponHolderSimulated_Scripted"])
+		};
+		_trackedUnits append _vehicles;
+	};
 
-		    		_isEmptyVehicle = _unit isKindOf "LandVehicle" && ({alive _x} count (crew _unit) == 0);
-					_isMan = (vehicle _unit) isKindOf "Man";
-					_isCustomObject = _unit getVariable ["GRAD_replay_track", false];
-					_isAI = !isPlayer _unit;
-					// _isPlayerVehicle = _unit isPlayer;
+	if (GRAD_REPLAY_AI_ONFOOT_TRACKED) then {
+		_ai = allUnits - playableUnits - switchableUnits;
+		_trackedUnits append _ai;
+	};
 
-		    		_name = if (alive _unit && _isMan) then {name _unit} else {""};
-		    		_groupname = if (_unit isEqualTo (leader group _unit)) then {" (" + groupId (group _unit) + ")"} else {""};
-					_veh = vehicle _unit;
-					_pos = getpos _unit;
-					_side = side _unit;
-					_color = [_side] call GRAD_replay_fnc_getSideColor;
-					_type = typeOf (vehicle _unit);
-					_icon = getText (configfile >> "CfgVehicles" >> _type >> "icon");
+	/* diag_log ["_trackedUnits",_trackedUnits]; */
 
+	// _nextTickData contains all the data of the next timestamp that is then appended to GRAD_REPLAY_DATABASE
+	_nextTickData = [];
 
-					// mark funkwagen if he is sending in red // speciality for mission "breaking contact"
-					if (_type isEqualTo "rhs_gaz66_r142_vv" && {_veh getVariable ["tf_range",0] == 50000}) then {
-						_color = [1,0,0,1];
-					};
+	{
+		_unit = _x;
+		_veh = vehicle _unit;
```

Comment:
`private` keyword?

Reply:
This is all running inside a CBA perFrameHandler. Does it need `private`?

Reply:
not really. But private also provides a small performance advantage as the engine would otherwise need to check all scopes above the current one about whether the variable is already defined.